### PR TITLE
Typehint common dunder methods

### PR DESCRIPTION
**Pull request by intgr**
Was suggested in #1206 comments.

### CODE REVIEW

File: django-stubs/contrib/gis/gdal/field.pyi [hunk 1]
```diff
@@ -4,7 +4,6 @@ from django.contrib.gis.gdal.base import GDALBase as GDALBase
 
 class Field(GDALBase):
     ptr: Any = ...
-    __class__: Any = ...
```
Comment (intgr) [inline django-stubs/contrib/gis/gdal/field.pyi:7]: These look like a mistake.

File: django-stubs/contrib/gis/db/backends/base/adapter.pyi [hunk 2]
```diff
@@ -4,5 +4,5 @@ class WKTAdapter:
     wkt: Any = ...
     srid: Any = ...
     def __init__(self, geom: Any) -> None: ...
-    def __eq__(self, other: Any) -> Any: ...
+    def __eq__(self, other: Any) -> bool: ...
     def __hash__(self) -> Any: ...
```
Comment (sobolevn) [inline django-stubs/contrib/gis/db/backends/base/adapter.pyi]: ` -> int`

File: django-stubs/contrib/gis/gdal/layer.pyi [hunk 3]
```diff
@@ -7,7 +7,7 @@ class Layer(GDALBase):
     def __init__(self, layer_ptr: Any, ds: Any) -> None: ...
     def __getitem__(self, index: Any) -> Any: ...
     def __iter__(self) -> Any: ...
```
Comment (sobolevn) [inline django-stubs/contrib/gis/gdal/layer.pyi]: ` -> Iterator[Any]`

File: django-stubs/contrib/gis/geos/geometry.pyi [hunk 4]
```diff
@@ -10,15 +10,14 @@ class GEOSGeometryBase(GEOSBase):
     ptr_type: Any = ...
     destructor: Any = ...
     has_cs: bool = ...
-    __class__: Any = ...
     def __init__(self, ptr: Any, cls: Any) -> None: ...
     def __copy__(self) -> Any: ...
```
Comment (sobolevn) [inline django-stubs/contrib/gis/geos/geometry.pyi]: `__copy__` and `__deepcopy__` can return `Self` type.

File: django-stubs/utils/functional.pyi [hunk 5]
```diff
@@ -65,7 +65,6 @@ class LazyObject:
     __bytes__: Callable = ...
     __bool__: Callable = ...
     __dir__: Callable = ...
-    __class__: Any = ...
     __ne__: Callable = ...
     __hash__: Callable = ...
```
Comment (sobolevn) [inline django-stubs/utils/functional.pyi:69]: This can be turned into methods in the next PRs :)

**Review by sobolevn — COMMENTED** [review]
Thank you! A couple of extra methods as ideas :)

File: django-stubs/contrib/gis/geos/geometry.pyi (hunk 4 above)
Comment (intgr) [reply django-stubs/contrib/gis/geos/geometry.pyi]: You mean `from typing_extensions import Self`?

Comment (intgr) [reply django-stubs/contrib/gis/geos/geometry.pyi]: Ah nope, got to use the `self: _T` trick still.

**Review by sobolevn — APPROVED** [review]
Thanks, great work!

**Comment by sobolevn** [comment]
You can now merge PRs yourself, go ahead :)

**Comment by intgr** [comment]
Still seeing this message

> Only those with [write access](https://docs.github.com/articles/what-are-the-different-access-permissions) to this repository can merge pull requests.

Tried logging out and back in.

**Comment by intgr** [comment]
@sobolevn In case you missed the previous comment: I still don't have permissions to merge.

**Comment by sobolevn** [comment]
Sorry, this is the first time I use teams inside an org inside a repo :(
I think I've messed something up.

But, it should be fine now! I've sent you a new invite.
Can you please double check? 

@adamchainz, it should be fine for you as well. Please, confirm :)

**Comment by sobolevn** [comment]
Oh, it is fine for @intgr 
Congrats 🎉

**Comment by intgr** [comment]
Yep it worked now, awesome!

"Squash and merge" is the correct flow in this repository, right? Maybe we can disable the other merge options in project settings? So I can't click the wrong button by accident :)

**Comment by sobolevn** [comment]
I also sometimes (rarely) use rebase and merge (when all commits make sense).

I will change it, thank you!

**Comment by adamchainz** [comment]
Nice work here, thanks.

**Pull request merged**